consoleme/handlers/v2/challenge.py

```diff
@@ -106,88 +107,116 @@ async def get(self, requested_challenge_token):
-                    self.write({"message": message})
-                return
-            user_challenge["status"] = "success"
-            user_challenge["user"] = self.user
-            user_challenge["groups"] = self.groups
-            red.hset(
-                config.get("challenge_url.redis_key", "TOKEN_CHALLENGES_TEMP"),
-                requested_challenge_token,
-                json.dumps(user_challenge),
+
+        request_ip = valid_user_challenge["ip"]
+        request_user = valid_user_challenge["user"]
+        message = (
+            f"A user at **{request_ip}** has requested ConsoleMe credentials for **{request_user}**.\n\n"
+            f"You must approve this request for credentials to be provided. "
+            f"You will not be able to refresh or revisit this page after closing it.\n\n"
+            f"If you did not create this request, please report it to your security team."
+        )
+
+        self.write(
+            {
+                "message": message,
+                "nonce": valid_user_challenge["nonce"],
+                "show_approve_button": True,
+            }
+        )
+
+    async def post(self, requested_challenge_token):
+        if not config.get("challenge_url.enabled", False):
+            raise MissingConfigurationValue(
+                "Challenge URL Authentication is not enabled in ConsoleMe's configuration"
             )
-            message = "You've successfully authenticated to ConsoleMe and may now close this page."
-            if endpoint == "web":
-                self.write(message)
-            elif endpoint == "api":
-                self.write({"message": message})
-        else:
-            message = "The requested challenge URL was not found. Please try requesting a new challenge URL."
-            if endpoint == "web":
-                self.write(message)
-            elif endpoint == "api":
-                self.write({"message": message})
+        data = tornado.escape.json_decode(self.request.body)
+        ip = self.request.headers.get("X-Forwarded-For", self.request.remote_ip).split(
+            ","
+        )
+        if isinstance(ip, list):
+            ip = ip[0]
+        log_data = {
+            "user": self.user,
+            "function": f"{__name__}.{self.__class__.__name__}.{sys._getframe().f_code.co_name}",
+            "requested_challenge_token": requested_challenge_token,
+            "message": "Incoming request",
+            "ip": ip,
+        }
+        log.debug(log_data)
+
+        all_challenges = red.hgetall(
+            config.get("challenge_url.redis_key", "TOKEN_CHALLENGES_TEMP")
+        )
+        if not all_challenges:
+            message = (
+                "Unable to find a matching challenge URL. This usually means that it has expired. "
+                "Please try requesting a new challenge URL."
+            )
+            self.write({"message": message})
             return
 
+        await delete_expired_challenges(all_challenges)
```

Comment:
Do we need to do this? It looks like we're setting a TTL on the keys, so Redis should take care of it for us.

Reply:
Redis hashe keys don't support a ttl :-(. only redis strings do, but then that muddies up the database, so we just handle the ttl ourselves